File: haskell/src/SpaceTraders/Database/Contracts.hs
{-# LANGUAGE OverloadedStrings #-}

module SpaceTraders.Database.Contracts
  ( getContracts
  , setContract
  ) where

import           Control.Monad.Error.Class
import           Control.Monad.Reader
import           Data.Aeson
import qualified Database.SQLite.Simple      as S

import           SpaceTraders
import           SpaceTraders.Model.Contract
import           SpaceTraders.Utils

getContracts :: (HasDatabaseConn env, MonadIO m, MonadReader env m) => m [Contract]
getContracts = query_ "SELECT data FROM contracts;"

setContract :: (HasDatabaseConn env, MonadError e m, MonadIO m, MonadReader env m) => Contract -> m ()
setContract contract = updateContract `catchError` addContract
  where
    addContract _ = execute "INSERT INTO contracts(data) VALUES (json(?));" (S.Only (encode contract))
    updateContract = execute "UPDATE contracts SET data = json(?) WHERE data->>'contractId' = ?;" (encode contract, contractId contract)

File: nodejs/database/contracts.js
import db from './db.js';

const addContractStatement = db.prepare(`INSERT INTO contracts(data) VALUES (json(?));`);
const getContractStatement = db.prepare(`SELECT data FROM contracts WHERE data->>'id' = ?;`);
const getContractsStatement = db.prepare(`SELECT data FROM contracts WHERE data->>'fulfilled' = false;`);
const updateContractStatement = db.prepare(`UPDATE contracts SET data = json(:data) WHERE data->>'id' = :id;`);

export function getContract(id) {
	const data = getContractStatement.get(id);
	if (data === undefined) {
		return null;
	}
	return JSON.parse(data.data);
}

export function getContracts() {
	const data = getContractsStatement.all();
	return data.map(contractData => JSON.parse(contractData.data));
}

export function setContract(data) {
	if (getContract(data.id) === null) {
		return addContractStatement.run(JSON.stringify(data)).lastInsertRowid;
	} else {
		return updateContractStatement.run({
			data: JSON.stringify(data),
			id: data.id,
		}).changes;
	}
}
